### PR TITLE
Configure retention and increase volume size

### PR DESCRIPTION
**Pull request by omegas27**

### CODE REVIEW

**Comment by omegas27** [comment]
Tested on gorgoth:
```
      --storage.tsdb.retention.size=45GB
      --config.file=/etc/prometheus/config_out/prometheus.env.yaml
      --storage.tsdb.path=/prometheus
      --storage.tsdb.retention.time=2w
      --web.enable-lifecycle
      --storage.tsdb.no-lockfile
      --web.external-url=https://prometheus.g8s.gorgoth.gridscale.kvm.gigantic.io/gorgoth
      --web.route-prefix=/gorgoth
      --storage.tsdb.wal-compression
```

File: helm/prometheus-meta-operator/templates/configmap.yaml [hunk 1]
```diff
@@ -24,9 +24,12 @@ data:
       prometheus:
         address: {{ .Values.Installation.V1.Monitoring.Prometheus.Address }}
         baseDomain: {{ .Values.Installation.V1.Monitoring.Prometheus.Host }}
+        retention:
+          duration: 2w
+          size: 90GB
         storage:
           createPVC: {{ and .Values.Installation.V1.Monitoring.Prometheus.Storage.Enabled .Values.Installation.V1.Monitoring.Prometheus.Storage.CreatePVC }}
-          size: 20Gi
```
Comment (TheoBrigitte) [inline helm/prometheus-meta-operator/templates/configmap.yaml:29]: Resizing pvc via statefulset is not supported yet in kubernetes: https://github.com/kubernetes/kubernetes/issues/68737

**Review by TheoBrigitte — CHANGES_REQUESTED** [review]

File: helm/prometheus-meta-operator/templates/configmap.yaml (hunk 1 above)
Comment (omegas27) [reply helm/prometheus-meta-operator/templates/configmap.yaml:29]: I am not sure what you want me to change here. We still need to increase it to 100Gi

Comment (TheoBrigitte) [reply helm/prometheus-meta-operator/templates/configmap.yaml:29]: I'm just saying this does not work for existing cluster. The volume size won't be changed. So we need a workaround.

Comment (omegas27) [reply helm/prometheus-meta-operator/templates/configmap.yaml:29]: So we would need a volume deleter if the volume size is 20Gi right?

Comment (TheoBrigitte) [reply helm/prometheus-meta-operator/templates/configmap.yaml:29]: Yeah right, official doc says you have to manually delete the pvc created by a stateful set when cleaning it (also valable for resizing): https://kubernetes.io/docs/tasks/run-application/delete-stateful-set/#complete-deletion-of-a-statefulset

File: helm/prometheus-meta-operator/templates/rbac.yaml [hunk 2]
```diff
@@ -49,6 +49,21 @@ rules:
     verbs:
       - create
       - delete
+  - apiGroups:
```
Comment (omegas27) [inline helm/prometheus-meta-operator/templates/rbac.yaml:52]: Can you add a todo here to say we need to remove it once the resize is done and not leave some leftover rights?

File: main.go [hunk 3]
```diff
@@ -112,7 +112,9 @@ func mainE(ctx context.Context) error {
 	daemonCommand.PersistentFlags().String(f.Service.Prometheus.BaseDomain, "", "Base domain to create Prometheus Ingress resources under.")
 	daemonCommand.PersistentFlags().String(f.Service.Prometheus.Address, "", "Address to access Prometheus UI.")
 	daemonCommand.PersistentFlags().Bool(f.Service.Prometheus.Storage.CreatePVC, false, "Should the operator create a PVC for storage.")
-	daemonCommand.PersistentFlags().String(f.Service.Prometheus.Storage.Size, "20Gi", "Storage size for prometheus.")
+	daemonCommand.PersistentFlags().String(f.Service.Prometheus.Storage.Size, "50Gi", "Storage size for prometheus.")
```
Comment (omegas27) [inline main.go:115]: Should this not be 100Gi and the retention size would be 90 as well?

File: service/controller/resource/volume-resize-hack/create.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
+package volumeresizehack
```
Comment (omegas27) [inline service/controller/resource/volume-resize-hack/create.go:1]: I like the direction of this

**Comment by omegas27** [comment]
I just commented a few things but this looks really nice

File: helm/prometheus-meta-operator/templates/rbac.yaml (hunk 2 above)
Comment (TheoBrigitte) [reply helm/prometheus-meta-operator/templates/rbac.yaml:52]: done

File: main.go (hunk 3 above)
Comment (TheoBrigitte) [reply main.go:115]: You're probably right, but I always got 100Gi during my tests.

Comment (TheoBrigitte) [reply main.go:115]: That's because we set this value in the helm chart.

File: helm/prometheus-meta-operator/templates/rbac.yaml (hunk 2 above)
Comment (omegas27) [reply helm/prometheus-meta-operator/templates/rbac.yaml:53]: :1st_place_medal:

File: main.go (hunk 3 above)
Comment (TheoBrigitte) [reply main.go]: done

**Comment by TheoBrigitte** [comment]
This was successfully tested. The workflow might seems strange and goes as follow:

1. We detect PVC exist and does not have the same size as defined in the StS
2. Delete the PVC
3. Scale down StS; so pod is deleted and volume is freed
4. Wait for PVC to be deleted
5. Scale down StS again; we need this because prometheus-operator scale back up the StS and the statefuleset-controller does not kick to re-create the PVC from StS volume claim template. Also trying to pause prometheus-operator reconciliation via the pause field results in the StS being deleted when trying to scale it down.
6. prometheus-operator scale back up the StS once more, statefulset-controller kicks in and re-create the PVC.

**Comment by omegas27** [comment]
I cannot approve but this looks legit to me

**Pull request merged**